src/coreComponents/constitutive/fluid/RachfordRice.cpp

```diff
@@ -0,0 +1,187 @@
+  xMin += sqrtEpsilon * ( LvArray::math::abs( xMin ) + sqrtEpsilon );
+  xMax -= sqrtEpsilon * ( LvArray::math::abs( xMax ) + sqrtEpsilon );
+
+  real64 currentError = 1 / epsilon;
+
+  // step 2: start the SSI loop
+  real64 funcXMin = 0.0;
+  real64 funcXMid = 0.0;
+  real64 funcXMax = 0.0;
+  bool recomputeMin = true;
+  bool recomputeMax = true;
+  integer SSIIteration = 0;
+
+  while( ( currentError > SSITolerance ) && ( SSIIteration < maxSSIIterations ) )
+  {
+    real64 const xMid = 0.5 * ( xMin + xMax );
+    if( recomputeMin )
+    {
+      funcXMin = evaluate( kValues, feed, presentComponentIds, xMin );
+    }
+    if( recomputeMax )
+    {
+      funcXMax = evaluate( kValues, feed, presentComponentIds, xMax );
+    }
+    funcXMid = evaluate( kValues, feed, presentComponentIds, xMid );
+
+    if( ( funcXMin < 0 ) && ( funcXMax < 0 ) )
+    {
+      return gasPhaseMoleFraction = 0.0;
+    }
+    else if( ( funcXMin > 1 ) && ( funcXMax > 1 ) )
+    {
+      return gasPhaseMoleFraction = 1.0;
+    }
+    else if( funcXMin * funcXMid < 0.0 )
+    {
+      xMax = xMid;
+      recomputeMax = true;
+      recomputeMin = false;
+    }
+    else if( funcXMax * funcXMid < 0.0 )
+    {
+      xMin = xMid;
+      recomputeMax = false;
+      recomputeMin = true;
+    }
+
+    currentError = LvArray::math::min( LvArray::math::abs( funcXMax - funcXMin ),
+                                       LvArray::math::abs( xMax - xMin ) );
+    SSIIteration++;
+
+    // TODO: add warning if max number of SSI iterations is reached
+  }
+
+  gasPhaseMoleFraction = 0.5 * ( xMax + xMin );
+
+  // step 3: start the Newton loop
+  integer newtonIteration = 0;
+  real64 newtonValue = gasPhaseMoleFraction;
+
+  while( ( currentError > newtonTolerance ) && ( newtonIteration < maxNewtonIterations ) )
+  {
+    real64 const deltaNewton = -evaluate( kValues, feed, presentComponentIds, newtonValue )
+                               / evaluateDerivative( kValues, feed, presentComponentIds, newtonValue );
+    currentError = LvArray::math::abs( deltaNewton ) / LvArray::math::abs( newtonValue );
+
+    // test if we are stepping out of the [xMin;xMax] interval
+    if( newtonValue + deltaNewton < xMin )
+    {
+      newtonValue = 0.5 * ( newtonValue + xMin );
+    }
+    else if( newtonValue + deltaNewton > xMax )
+    {
+      newtonValue = 0.5 * ( newtonValue + xMax );
+    }
+    else
+    {
+      newtonValue = newtonValue + deltaNewton;
+    }
+    newtonIteration++;
+
+    // TODO: add warning if max number of Newton iterations is reached
+  }
+  return gasPhaseMoleFraction = newtonValue;
+}
+
+real64
+RachfordRice::evaluate( arrayView1d< real64 const > const kValues,
+                        arrayView1d< real64 const > const feed,
+                        arrayView1d< integer const > const presentComponentIds,
+                        real64 const & x )
+{
+  real64 value = 0.0;
+  for( integer ic = 0; ic < presentComponentIds.size(); ++ic )
```

Comment:
Should be `for (integer const ic : presentComponentIds)` i.e. loop over the actual values of the indicator array.

Reply:
Thanks for the catch, this is fixed now